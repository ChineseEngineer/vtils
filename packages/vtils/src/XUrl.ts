/**
 * 将 File 转换为字符串。
 */
export class XUrl {
  private static divider = '\n=^=^=^=^=\n'

  static extractUrl<TUrl>(url: TUrl): TUrl {
    return (
      typeof url !== 'string'
        ? url
        : url.split(XUrl.divider)[0]
    ) as any
  }

  static extractFile(url: any): Promise<File | undefined> {
    return new Promise(resolve => {
      if (typeof url !== 'string') return resolve()

      const [fileUrl, fileStrInfo]: [string, string | undefined] = url.split(XUrl.divider) as any

      if (!fileStrInfo) return resolve()

      let fileInfo: File
      try {
        fileInfo = JSON.parse(fileStrInfo) as File
      } catch (e) {
        /* istanbul ignore next */
        return resolve()
      }

      const xhr = new XMLHttpRequest()
      xhr.open('GET', fileUrl)
      xhr.responseType = 'blob'
      xhr.onload = () => {
        const file = new File([xhr.response], fileInfo.name, fileInfo)
        resolve(file)
      }
      xhr.onerror = () => resolve()
      xhr.send()
    })
  }

  static extract<TUrl>(url: TUrl): Promise<{ url: TUrl, file: File | undefined }> {
    return Promise
      .all([
        Promise.resolve(XUrl.extractUrl(url)),
        XUrl.extractFile(url),
      ])
      .then(([url, file]) => ({url, file}))
  }

  static generate(file: File): string {
    return `${URL.createObjectURL(file)}${XUrl.divider}${JSON.stringify({
      name: file.name,
      size: file.size,
      type: file.type,
      lastModified: file.lastModified,
    } as Partial<File>)}`
  }
}
