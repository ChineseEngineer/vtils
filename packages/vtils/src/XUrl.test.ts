import http from 'http'
import {XUrl} from './XUrl'

async function fileToObj(file: File): Promise<Partial<File> & { content: string }> {
  const fileContent = await new Promise<string>(resolve => {
    const fileReader = new FileReader()
    fileReader.readAsText(file)
    fileReader.onload = () => {
      resolve(fileReader.result as any)
    }
  })
  return {
    content: fileContent,
    name: file.name,
    type: file.type,
    size: file.size,
    lastModified: file.lastModified,
  }
}

describe('XUrl', () => {
  let server: http.Server
  let testFile: File

  beforeAll(() => {
    window.URL.createObjectURL = jest.fn().mockImplementation(() => 'http://localhost:56384')
    server = http
      .createServer((_, res) => {
        res.setHeader('access-control-allow-origin', '*')
        res.end('hello')
      })
      .listen(56384)
    testFile = new File(
      ['hello'],
      'test.txt',
      {
        type: 'text/plain',
        lastModified: 1574155444339,
      },
    )
  })

  afterAll(() => {
    (window.URL.createObjectURL as any).mockReset()
    server.close()
  })

  test('generate', () => {
    const xUrl = XUrl.generate(testFile)
    expect(xUrl).toMatchSnapshot()
  })

  test('extractUrl', () => {
    const xUrl = XUrl.generate(testFile)
    const url = XUrl.extractUrl(xUrl)
    expect(url).toMatchSnapshot()
  })

  test('extractFile', async () => {
    const xUrl = XUrl.generate(testFile)
    const file = (await XUrl.extractFile(xUrl))!
    expect(await fileToObj(file)).toMatchSnapshot()
  })

  test('extract', async () => {
    const xUrl = XUrl.generate(testFile)
    const res = await XUrl.extract(xUrl)
    expect({
      ...res,
      file: await fileToObj(res.file!),
    }).toMatchSnapshot()
  })

  test('正常的 URL', async () => {
    expect(await XUrl.extract('http://foo.bar')).toMatchSnapshot()
  })

  test('自定义的 URL', async () => {
    expect(await XUrl.extract(['hello'])).toMatchSnapshot()
  })

  test('transform', async () => {
    const xUrl = XUrl.generate(testFile)
    const data = {
      id: 1,
      title: 'hello',
      icon: xUrl,
      users: [
        {
          id: 2,
          avatar: xUrl,
        },
        {
          id: 3,
          avatar: xUrl,
        },
        {
          id: 4,
          avatar: xUrl,
          albums: [xUrl, xUrl, xUrl],
        },
      ],
    } as const
    expect(await XUrl.transform(data, async () => 'url://ok')).toMatchSnapshot()
  })
})
