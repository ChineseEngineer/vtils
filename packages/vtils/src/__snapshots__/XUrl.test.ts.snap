// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XUrl extract 1`] = `
Object {
  "file": Object {
    "content": "hello",
    "lastModified": 1574155444339,
    "name": "test.txt",
    "size": 5,
    "type": "text/plain",
  },
  "url": "http://localhost:56384",
}
`;

exports[`XUrl extractFile 1`] = `
Object {
  "content": "hello",
  "lastModified": 1574155444339,
  "name": "test.txt",
  "size": 5,
  "type": "text/plain",
}
`;

exports[`XUrl extractUrl 1`] = `"http://localhost:56384"`;

exports[`XUrl generate 1`] = `
"٦ᄑXUrl㌳ঙ
http://localhost:56384
{\\"name\\":\\"test.txt\\",\\"size\\":5,\\"type\\":\\"text/plain\\",\\"lastModified\\":1574155444339}"
`;

exports[`XUrl transform 1`] = `
Object {
  "icon": "url://ok",
  "id": 1,
  "title": "hello",
  "users": Array [
    Object {
      "avatar": "url://ok",
      "id": 2,
    },
    Object {
      "avatar": "url://ok",
      "id": 3,
    },
    Object {
      "albums": Array [
        "url://ok",
        "url://ok",
        "url://ok",
      ],
      "avatar": "url://ok",
      "id": 4,
    },
  ],
}
`;

exports[`XUrl 正常的 URL 1`] = `
Object {
  "file": undefined,
  "url": "http://foo.bar",
}
`;

exports[`XUrl 自定义的 URL 1`] = `
Object {
  "file": undefined,
  "url": Array [
    "hello",
  ],
}
`;
